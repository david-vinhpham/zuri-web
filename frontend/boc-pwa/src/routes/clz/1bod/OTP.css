/**
 * Home Credit (https://www.homecredit.vn/)
 *
 * Copyright Â© 2018-present HomeCredit, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

@import '../../../components/variables.css';

:root {
  --popup-text-line-height: 19.25px;
}

.title {
  text-align: center;
  color: var(--c-text);
  font-size: var(--f-text);
  padding-top: var(--b-space-lg);
  padding-bottom: var(--b-space-xs);
  margin-top: 0;
  font-style: italic;
}

.otpWrapper {
  height: var(--b-space-5xl);
}

.modalFooter {
  padding: var(--b-space-3xs) var(--b-space-md);
  background-color: var(--c-brand);
  color: white;
  text-align: center;
}

.resendBody {
  text-align: center;
  color: var(--c-text);
  font-size: var(--f-text);
  font-style: italic;
  padding-bottom: var(--b-space-md);

  & h3 {
    font-style: normal;
  }
}

.resendText {
  color: var(--c-brand);
}

.checkBox span,
a {
  color: var(--c-text);
  font-size: var(--f-text);
}

.checkBox a {
  text-decoration: underline;
  font-style: italic;
}

.checkBox {
  display: block !important;
}

.termText {
  text-align: left;
  white-space: normal;
  line-height: var(--popup-text-line-height);
}

@media (--sm) {
  .checkBox {
    text-align: left;
  }
}
