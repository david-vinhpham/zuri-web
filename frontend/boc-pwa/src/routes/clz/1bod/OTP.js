/**
 * BOC VN (http://www.bocvietnam.com/)
 *
 * Copyright Â© 2018-present BOCVN, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React, { Component, Fragment } from 'react';
import { compose } from 'redux';
import {
  objectOf,
  any,
  func,
  object,
  bool,
  oneOfType,
  number,
} from 'prop-types';
import { connect } from 'react-redux';
import { withFormik } from 'formik';
import { get, pick, isEmpty, noop } from 'lodash';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import { Checkbox, NumberInput } from 'homecredit-ui';
import OtpCountDown from 'components/common/OTPCountDown';
import Modal from 'components/common/Modal';

import { getValidationSchema } from 'utils/validations';

import { verifyOTP, generateOTP, trackingInfo } from 'actions/clz';
import { FORM_STATUS, LOADER_STATUS } from 'constants/common';
import { sendGAEvent } from 'utils/ga';

import {
  INPUT_OTP_FIELDS,
  OTP_TEXTS,
  ACCEPTED_TERM_FIELDS,
  MAX_ATTEMPT,
  MAX_RETRY,
  STEPS,
  STEP_NAME,
  OTP_COUNT_SECOND,
} from './constant';
import s from './OTP.css';
import { mapStateToProps, mapDispatchToProps } from './commonProps';

class OTP extends Component {
  static propTypes = {
    formStatus: number,
    errors: objectOf(any),
    values: objectOf(any),
    setFieldValue: func,
    setFieldTouched: func.isRequired,
    onChange1BODInfo: func,
    updateFormStatus: func,
    validateForm: func,
    handleBlur: func,
    goToNextStep: func,
    getOTPSuccess: func,
    updateLoadingStatus: func,
    updateStep: func,
    touched: oneOfType([bool, object]),
    globalPopup: objectOf(any),
    firstBODInfo: objectOf(any),
    getTrackingToken: func.isRequired,
  };

  static defaultProps = {
    formStatus: FORM_STATUS.INITIAL,
    errors: {},
    values: {},
    setFieldValue: noop,
    onChange1BODInfo: noop,
    handleBlur: noop,
    updateFormStatus: noop,
    validateForm: noop,
    goToNextStep: noop,
    getOTPSuccess: noop,
    updateLoadingStatus: noop,
    updateStep: noop,
    touched: {},
    globalPopup: {},
    firstBODInfo: {},
  };

  constructor(props) {
    super(props);
    this.state = {
      isOpenModal: false,
      isOpenResendModal: false,
      isOpenAttempModal: false,
      isOTPExpire: false,
      attempt: 0,
      retry: 0,
    };
  }

  /**
   * Update Form Status to initial after render
   *
   * @return {void}
   */
  componentDidMount() {
    this.props.updateFormStatus(FORM_STATUS.INITIAL);

    sendGAEvent('Visit_input_OTP', 'Input_OTP', 'Input OTP');
  }

  /**
   * Validate when form status is changed
   *
   * @return {void}
   */
  async componentWillReceiveProps(nextProps) {
    const {
      updateFormStatus,
      formStatus,
      validateForm,
      setFieldTouched,
      updateLoadingStatus,
      values,
    } = this.props;
    const { formStatus: nextStatus } = nextProps;
    const { isOTPExpire } = this.state;

    if (formStatus === FORM_STATUS.INITIAL && formStatus !== nextStatus) {
      const errors = await validateForm();
      // Track forget to click consent
      !values.acceptedTerm && sendGAEvent('Agree_TnC', 'TnC', 'Input OTP');

      // Show error when OTP is expired, skip validate otp
      if (isOTPExpire) {
        updateFormStatus(FORM_STATUS.INITIAL);
        this.setState({ isOpenAttempModal: true });
        return;
      }

      if (isEmpty(errors)) {
        // Show loading
        updateLoadingStatus(LOADER_STATUS.ON);
        // Calling verify OTP API
        this.verifyOTP();
      } else {
        updateFormStatus(FORM_STATUS.INITIAL);
        Object.keys(errors).forEach(error => setFieldTouched(error));
      }
    }
  }

  /**
   * Clear valud in redux state before unmount
   *
   * @return {void}
   */
  componentWillUnmount() {
    this.updateFormInfo(INPUT_OTP_FIELDS.name, '');
    this.updateFormInfo(ACCEPTED_TERM_FIELDS.name, false);
  }

  /**
   * Update Form Field Checkbox
   *
   * @return {void}
   */
  onCheckedChange = value => {
    this.updateFormInfo(ACCEPTED_TERM_FIELDS.name, value);
  };

  /**
   * Update state when modal is opened
   */
  onOpenModal = () => {
    this.setState({
      isOpenModal: true,
    });

    sendGAEvent('View_TnC', 'TnC', 'Input OTP');
  };

  /**
   * Update state when modal is closed
   */
  onCloseModal = () => {
    this.setState({
      isOpenModal: false,
    });
  };

  /**
   * Update state when modal is opened
   */
  onOpenResendModal = () => {
    const {
      values,
      getOTPSuccess,
      globalPopup,
      updateLoadingStatus,
    } = this.props;

    sendGAEvent('Regenerate_OTP', 'OTP', 'Input OTP');

    updateLoadingStatus(LOADER_STATUS.ON);
    generateOTP(values.phoneNumber).then(response => {
      updateLoadingStatus(LOADER_STATUS.OFF);
      if (isEmpty(response.errors)) {
        getOTPSuccess(response);
        this.setState({
          isOpenResendModal: true,
          retry: this.state.retry + 1,
          attempt: 0,
          isOTPExpire: false,
        });
      } else {
        globalPopup.open();
      }
    });
  };

  /**
   * Update state when modal is closed
   */
  onCloseResendModal = () => {
    this.isMaxRetry
      ? this.props.updateStep(STEPS.INPUT_PHONE)
      : this.setState({
          isOpenResendModal: false,
        });
  };

  /**
   * Update state when modal is opened
   */
  onOpenAttempModal = () => {
    sendGAEvent('Incorrect_message', 'OTP', 'Input OTP');

    if (!this.state.isOpenAttempModal) {
      this.props.updateFormStatus(FORM_STATUS.INITIAL);
      this.setState({
        isOpenAttempModal: true,
        attempt: this.state.attempt + 1,
      });
    }
  };

  /**
   * Update state when modal is closed
   */
  onCloseAttempModal = () => {
    if (this.isMaxAttempt) {
      this.props.updateStep(STEPS.INPUT_PHONE);
    } else {
      this.setState({
        isOpenAttempModal: false,
      });
    }
  };

  /**
   * Update Form Field Input
   *
   * @return {void}
   */
  onFieldChange = (name, value) => {
    this.updateFormInfo(name, value);
  };

  get isMaxRetry() {
    return MAX_RETRY - this.state.retry <= 0;
  }

  get isMaxAttempt() {
    return MAX_ATTEMPT - this.state.attempt <= 0;
  }

  /**
   * Call API - Verify OTP
   *
   * @return {void}
   */
  verifyOTP = () => {
    const {
      goToNextStep,
      globalPopup,
      updateLoadingStatus,
      firstBODInfo,
      getTrackingToken,
    } = this.props;
    const { phoneNumber, otp, verificationID } = firstBODInfo;

    verifyOTP(phoneNumber, verificationID, otp).then(response => {
      updateLoadingStatus(LOADER_STATUS.OFF);
      if (isEmpty(response.errors)) {
        const isMatch = get(response, 'data.status');
        if (isMatch === 'MATCHED') {
          goToNextStep();
          const {
            creditAmount,
            bestOffer,
            retailAgent,
            sa,
            tenor,
            partnerCode,
            tipperCode,
          } = firstBODInfo;
          const annualInterestRate = get(
            bestOffer,
            'celOffer.annualInterestRate',
          );
          const firstInstallmentDueDate = get(
            bestOffer,
            'celOffer.firstInstallmentDueDate',
          );
          const offerCode = get(bestOffer, 'offerCode');
          const productCode = get(bestOffer, 'productInformation.code');
          const productType = get(bestOffer, 'productInformation.type');
          const monthlyPayment = get(
            bestOffer,
            'celOffer.monthlyInstallmentAmount.amount',
          );
          const trackingToken = get(response, 'token.token');
          const data = {
            url: STEP_NAME[STEPS.CALCULATE_CASH],
            token: trackingToken,
            body: {
              annualInterestRate,
              creditAmount,
              firstInstallmentDueDate,
              monthlyPayment,
              offerCode,
              productCode,
              productType,
              retailAgent,
              sa,
              tenor,
              partnerCode,
              tipperCode,
            },
          };
          trackingInfo(data);
          getTrackingToken(trackingToken);
        } else {
          this.onOpenAttempModal();
        }
      } else {
        globalPopup.open();
      }
    });
  };

  /**
   * Update Form Field Value & Redux State
   *
   * @param  {String} name  Field Name
   * @param  {String} value Field Value
   *
   * @return {Void}
   */
  updateFormInfo = (name, value) => {
    const { setFieldValue, onChange1BODInfo } = this.props;

    // Update Form Value
    setFieldValue(name, value);

    // Update Redux state
    onChange1BODInfo(name, value);
  };

  /**
   * Hanlde time out OTP
   */
  handleTimeoutOTP = () => {
    this.setState({
      isOTPExpire: true,
    });
  };

  handleBlur = e => {
    sendGAEvent('Input_OTP', 'Input_OTP', 'Input OTP');

    this.props.handleBlur(e);
  };

  render() {
    const { values, touched, errors } = this.props;
    const {
      isOpenModal,
      isOpenResendModal,
      isOpenAttempModal,
      attempt,
      retry,
      isOTPExpire,
    } = this.state;
    const { name, type, label } = INPUT_OTP_FIELDS;
    const otp = get(values, name);
    const phoneNumber = !isEmpty(values.phoneNumber)
      ? values.phoneNumber.replace(/(\d{4})\d{3}/, '$1xxx')
      : '';
    const attempRemain = MAX_ATTEMPT - attempt;
    const resendRemain =
      MAX_RETRY - retry > 1 ? `0${MAX_RETRY - retry - 1}` : '0';

    return (
      <div className="container">
        <div className="row center">
          <div className="col-md-10">
            <p className={s.title}>
              {`${OTP_TEXTS.DESCRIPTION}`}
              <b>{`${phoneNumber}`}</b>
            </p>
            <div className={s.otpWrapper}>
              <NumberInput
                autoFocus
                id={name}
                key={name}
                type={type}
                label={label}
                name={name}
                value={otp}
                onChange={this.onFieldChange}
                onBlur={this.handleBlur}
                touched={touched[name]}
                errorMsg={errors[name]}
                maxLength={8}
                disabled={isOTPExpire}
                msgPosition="below"
              />
            </div>
            <OtpCountDown
              start={!isOTPExpire}
              disabled={isOTPExpire}
              seconds={OTP_COUNT_SECOND}
              setOTPTimeOut={this.handleTimeoutOTP}
              resendOTP={this.onOpenResendModal}
              retry={MAX_RETRY}
            />
            <Checkbox
              className={s.checkBox}
              checked={values.acceptedTerm}
              label={OTP_TEXTS.TERM}
              onChange={this.onCheckedChange}
              error={!!errors.acceptedTerm && touched.acceptedTerm}
            >
              <Modal
                onOpen={this.onOpenModal}
                onClose={this.onCloseModal}
                isOpen={isOpenModal}
                toggleButton={
                  <a>
                    <b>{OTP_TEXTS.CONDITION}</b>
                  </a>
                }
                footer={
                  <div className={s.modalFooter} onClick={this.onCloseModal}>
                    <h4>{OTP_TEXTS.AGREE}</h4>
                  </div>
                }
              >
                <div className={s.resendBody}>
                  <h3>Äiá»u khoáº£n vÃ  Äiá»u kiá»n</h3>
                  <p className={s.termText}>{OTP_TEXTS.TERM_CONDITION}</p>
                </div>
              </Modal>
            </Checkbox>
            <Modal
              onOpen={this.onOpenResendModal}
              onClose={this.onCloseResendModal}
              isOpen={isOpenResendModal}
              footer={
                <div
                  className={s.modalFooter}
                  onClick={this.onCloseResendModal}
                >
                  <h4>{OTP_TEXTS.AGREE}</h4>
                </div>
              }
            >
              <div className={s.resendBody}>
                <h3>LÆ°u Ã½</h3>
                {this.isMaxRetry ? (
                  <Fragment>
                    <p>{OTP_TEXTS.MAXIMUM_TRIED_1}</p>
                    <p>{OTP_TEXTS.MAXIMUM_TRIED_2}</p>
                  </Fragment>
                ) : (
                  <Fragment>
                    Báº¡n cÃ²n <span className={s.resendText}>{resendRemain}</span>{' '}
                    láº§n Gá»­i láº¡i MÃ£ XÃ¡c Thá»±c
                  </Fragment>
                )}
              </div>
            </Modal>
            <Modal
              onOpen={this.onOpenAttempModal}
              onClose={this.onCloseAttempModal}
              isOpen={isOpenAttempModal}
              footer={
                <div
                  className={s.modalFooter}
                  onClick={this.onCloseAttempModal}
                >
                  <h4>{OTP_TEXTS.AGREE}</h4>
                </div>
              }
            >
              <div className={s.resendBody}>
                <h3>LÆ°u Ã½</h3>
                {isOTPExpire &&
                  attempt < MAX_ATTEMPT && <p>{OTP_TEXTS.OTP_DISABLED}</p>}
                {(!isOTPExpire || (isOTPExpire && attempt === MAX_ATTEMPT)) && (
                  <AttempContent
                    isMaxAttempt={this.isMaxAttempt}
                    attempRemain={attempRemain}
                  />
                )}
              </div>
            </Modal>
          </div>
        </div>
      </div>
    );
  }
}

const AttempContent = ({ isMaxAttempt, attempRemain }) =>
  isMaxAttempt ? (
    <Fragment>
      <p>
        {OTP_TEXTS.MAXIMUM_ATTEMP_1}
        <span className={s.resendText}>{`0${MAX_ATTEMPT}`}</span>
        {OTP_TEXTS.MAXIMUM_ATTEMP_2}
      </p>
      <p>{OTP_TEXTS.MAXIMUM_ATTEMP_3}</p>
    </Fragment>
  ) : (
    <Fragment>
      <p>{OTP_TEXTS.ATTEMP_1}</p>
      <p>
        {OTP_TEXTS.ATTEMP_2}
        <span className={s.resendText}>{`0${attempRemain}`}</span>
        {OTP_TEXTS.ATTEMP_3}
      </p>
    </Fragment>
  );
AttempContent.propTypes = {
  isMaxAttempt: bool.isRequired,
  attempRemain: number.isRequired,
};

const enhance = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  withFormik({
    validateOnBlur: false,
    mapPropsToValues: props =>
      pick(props.firstBODInfo, [
        'phoneNumber',
        'otp',
        'acceptedTerm',
        'verificationID',
      ]),
    validationSchema: getValidationSchema(['otp', 'acceptedTerm']),
  }),
  withStyles(s),
);

export default enhance(OTP);
