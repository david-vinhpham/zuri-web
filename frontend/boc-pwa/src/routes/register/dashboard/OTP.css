.verifyInput {
  padding-bottom: var(--b-space-lg);
}

.btnSendOTP {
  display: block;
  margin-left: auto;
  margin-right: auto;
  margin-top: var(--b-space-sm);
}
