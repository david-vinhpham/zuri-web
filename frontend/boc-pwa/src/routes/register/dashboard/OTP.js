import React from 'react';
import OtpCountDown from 'components/common/OTPCountDown';
import { Button, NumberInput } from 'homecredit-ui';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import { compose } from 'redux';
import { withFormik } from 'formik';
import { getValidationSchema } from 'utils/validations';
import isEmpty from 'lodash/isEmpty';
import { func, objectOf, any } from 'prop-types';
import noop from 'lodash/noop';

import s from './OTP.css';
import {
  PHONE_PALCEHOLDER,
  MAX_RETRY,
  SECOND_COUNTDOWN,
  MAX_TIME_RESEND_OTP,
} from './constant';

const name = 'otp';

class OTP extends React.PureComponent {
  state = {
    isOTPExpire: false,
    sendOTPCount: 1,
    retryVerify: 0,
  };

  handleTimeoutOTP = () => {
    this.setState({
      isOTPExpire: true,
    });
  };

  /**
   * @param {Boolean} isOverRetry
   */
  handleResendOTP = isOverRetry => {
    const { sendOTPCount } = this.state;
    const { overTry, reSendOTP } = this.props;
    const resendRemain =
      MAX_TIME_RESEND_OTP - sendOTPCount > 1
        ? `0${MAX_TIME_RESEND_OTP - sendOTPCount - 1}`
        : '0';

    if (isOverRetry) {
      overTry();
      return;
    }

    this.setState(() => ({
      isOTPExpire: false,
      sendOTPCount: sendOTPCount + 1,
      retryVerify: 0,
    }));

    reSendOTP(resendRemain);
  };

  handleVerifyOTP = async () => {
    const { retryVerify } = this.state;
    const {
      setFieldTouched,
      validateForm,
      registerDevice,
      values,
    } = this.props;

    setFieldTouched(name);
    const erros = await validateForm();
    if (isEmpty(erros)) {
      registerDevice(values[name], MAX_RETRY - (retryVerify + 1));

      this.setState({
        retryVerify: retryVerify + 1,
      });
    }
  };

  render() {
    const { isOTPExpire } = this.state;
    const { handleBlur, values, touched, errors, setFieldValue } = this.props;

    return (
      <div>
        <div className={s.verifyInput}>
          <NumberInput
            disabled={isOTPExpire}
            onChange={setFieldValue}
            onBlur={handleBlur}
            value={values[name]}
            label={PHONE_PALCEHOLDER}
            name={name}
            touched={touched[name]}
            msgPosition="below"
            errorMsg={errors[name]}
            maxLength={8}
          />
        </div>

        <OtpCountDown
          start={!isOTPExpire}
          disabled={isOTPExpire}
          setOTPTimeOut={this.handleTimeoutOTP}
          resendOTP={this.handleResendOTP}
          seconds={SECOND_COUNTDOWN}
          retry={MAX_TIME_RESEND_OTP}
        />

        <Button className={s.btnSendOTP} onClick={this.handleVerifyOTP}>
          Đăng ký thiết bị
        </Button>
      </div>
    );
  }
}

OTP.propTypes = {
  setFieldValue: noop,
  handleBlur: noop,
  values: objectOf(any),
  touched: objectOf(any),
  errors: objectOf(any),
  setFieldTouched: func,
  validateForm: func,
  registerDevice: func,
  overTry: func,
  reSendOTP: func,
};

OTP.defaultProps = {
  setFieldValue: noop,
  handleBlur: noop,
  values: {},
  touched: {},
  errors: {},
  setFieldTouched: noop,
  validateForm: noop,
  registerDevice: noop,
  overTry: noop,
  reSendOTP: noop,
};

const enchane = compose(
  withFormik({
    mapPropsToValues: () => ({ [name]: '' }),
    validationSchema: getValidationSchema([name]),
  }),
  withStyles(s),
);

export default enchane(OTP);
