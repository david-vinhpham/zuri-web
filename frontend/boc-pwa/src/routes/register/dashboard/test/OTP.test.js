import React from 'react';
import renderer from 'react-test-renderer';

import OTP from '../OTP';

it('OTP snapshot', () => {
  const wrapper = renderer.create(
    <OTP createAt="" unRegister={() => {}} showUnRegister />,
  );
  const snapshot = wrapper.toJSON();

  expect(snapshot).toMatchSnapshot();
});
