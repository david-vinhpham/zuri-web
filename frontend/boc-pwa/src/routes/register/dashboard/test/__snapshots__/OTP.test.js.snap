// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OTP snapshot 1`] = `
<div>
  <div
    className="verifyInput"
  >
    <div
      className="Input_wrapper__WALTg"
    >
      <i
        className="material-icons Input_icon__1jjsS"
      >
        
      </i>
      <input
        accept=""
        autoComplete=""
        className="Input_input__1IXXH"
        disabled={false}
        id=""
        maxLength={null}
        name="otp"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        placeholder=""
        readOnly={false}
        type="text"
        value=""
      />
      <label
        className="Input_label__2ZxWx"
        htmlFor=""
      >
        Nhập mã xác thực (OTP)
      </label>
      <div
        className="Input_underline__Xwqb3"
      />
    </div>
  </div>
  <div
    className="OTPCountDown"
  >
    <div
      className="timerLabel"
    >
      Mã xác thực sẽ hết hạn trong vòng
    </div>
    <div
      className="timer"
    >
      <span>
        05:00
      </span>
    </div>
  </div>
  <button
    className="Button_buttonReset__3cQTl Button_button__2pQRN btnSendOTP"
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    Đăng ký thiết bị
  </button>
</div>
`;
